src/config/GraphQLConfig.js

```diff
@@ -59,109 +64,143 @@ export async function getGraphQLConfig(configDir: Uri): Promise<GraphQLRC> {
+  // for the situation with many apps with a shared code in one repository.
+  // This function searches for the additional app configurations and
+  // returns the name of the app configuration if found.
+  getAppConfigNameByFilePath(filePath: Uri): ?string {
+    const appConfigs = this._config[APP_EXTENSIONS_NAME];
+    if (!appConfigs) {
+      return null;
     }
-    return config;
-  }
 
-  getConfigByFilePath(filePath: Uri): ?GraphQLConfig {
-    const name = this.getConfigNames().find(configName =>
-      this.getConfig(configName).isFileInInputDirs(filePath),
-    );
+    const appConfigNames = Object.keys(appConfigs);
 
-    return name ? this._configs[name] : null;
-  }
-}
-
-export class GraphQLConfig {
-  _config: Object;
-  _name: string;
-  _rootDir: Uri;
+    const name = appConfigNames.find(appName => {
+      const appConfig = this._config[APP_EXTENSIONS_NAME] &&
+        this._config[APP_EXTENSIONS_NAME][appName];
+      if (appConfig) {
+        return this.isFileInInputDirs(filePath, appName);
+      }
+      return false;
+    });
 
-  constructor(name: string, config: Object, rootDir: Uri): void {
-    this._name = name;
-    this._rootDir = rootDir;
-    this._config = config;
+    return name || null;
   }
 
   getRootDir(): Uri {
     return this._rootDir;
   }
 
-  getName(): string {
-    return this._name;
-  }
-
-  getConfig(): Object {
+  getConfig(): GraphQLConfiguration {
     return this._config;
   }
 
-  getInputDirs(): Array<string> {
-    return this._config['input-dirs'] ? this._config['input-dirs'] : [];
-  }
+  /**
+   * Below getters inspect the GraphQL configuration in two steps:
+   * 1. If `appName` is provided, look for the app configuration and try
+   *    returning app-specific properties, overriding the configuration options
+   *    at the top level.
+   * 2. If step 1 fails and/or `appName` is not passed in, look for the property
+   *    at the top level (the "root" configuration).
+   */
+
+  getInputDirs(appName: ?string): Array<Uri> {
+    if (appName) {
+      if (
+        this._config[APP_EXTENSIONS_NAME] &&
+        this._config[APP_EXTENSIONS_NAME][appName] &&
+        this._config[APP_EXTENSIONS_NAME][appName].inputDirs
+      ) {
+        return this._config[APP_EXTENSIONS_NAME][appName].inputDirs;
+      }
+    }
 
-  getExcludeDirs(): Array<string> {
-    return this._config['exclude-dirs'] ? this._config['exclude-dirs'] : [];
+    return this._config.inputDirs || [];
   }
 
-  isFileInInputDirs(fileName: string): boolean {
-    if (!this.getInputDirs()) {
-      return false;
+  getExcludeDirs(appName: ?string): Array<Uri> {
+    if (appName) {
+      if (
+        this._config[APP_EXTENSIONS_NAME] &&
+        this._config[APP_EXTENSIONS_NAME][appName] &&
+        this._config[APP_EXTENSIONS_NAME][appName].excludeDirs
+      ) {
+        return this._config[APP_EXTENSIONS_NAME][appName].excludeDirs;
+      }
     }
```

Comment:
Probably worth extracting this pattern into a separate function.

```
getConfig<T: any>(keyName: string, appName: ?string, defaultValue: T = []): T {
  if (
    appName &&
    this._config[APP_EXTENSIONS_NAME] &&
    this._config[APP_EXTENSIONS_NAME][appName] &&
    this._config[APP_EXTENSIONS_NAME][appName][keyName]
  ) {
    return this._config[APP_EXTENSIONS_NAME][appName][keyName];
  }
  return this._config[keyName] || defaultValue;
}
```

And then using it to implement `getExcludeDirs` and `getIncludeDirs` etc.